clients/drcachesim/tests/cache_replacement_policy_unit_test.cpp

```diff
@@ -49,24 +50,90 @@ class cache_lru_test_t : public cache_lru_t {
         }
     }
 
-    int
-    get_block_index(const addr_t addr)
+    void
+    access_and_check_lru(const addr_t addr,
+                         const int expected_replacement_way_after_access)
     {
-        addr_t tag = compute_tag(addr);
-        int block_idx = compute_block_idx(tag);
-        return block_idx;
+        memref_t ref;
+        ref.data.type = TRACE_TYPE_READ;
+        ref.data.size = 1;
+        ref.data.addr = addr;
+        request(ref);
+        assert(get_next_way_to_replace(get_block_index(addr)) ==
+               expected_replacement_way_after_access);
     }
+};
+
+template <class T> class cache_policy_test_t : public T {
+    int associativity_;
+    int line_size_;
+    int total_size_;
 
+public:
+    cache_policy_test_t(int associativity, int line_size, int total_size)
+    {
+        associativity_ = associativity;
+        line_size_ = line_size;
+        total_size_ = total_size;
+    }
     void
-    access_and_check_lru(const addr_t addr,
-                         const int expected_replacement_way_after_access)
+    initialize_cache()
+    {
+        caching_device_stats_t *stats = new cache_stats_t(line_size_, "", true);
+        if (!this->init(associativity_, line_size_, total_size_, nullptr, stats,
+                        nullptr)) {
+            std::cerr << "FIFO cache failed to initialize\n";
+            exit(1);
+        }
+    }
+
+    void
+    access_and_check_cache(const addr_t addr,
+                           const int expected_replacement_way_after_access)
+    {
+        memref_t ref;
+        ref.data.type = TRACE_TYPE_READ;
+        ref.data.size = 1;
+        ref.data.addr = addr;
+        this->request(ref);
+        assert(this->get_next_way_to_replace(this->get_block_index(addr)) ==
+               expected_replacement_way_after_access);
+    }
+
+    bool
+    tags_are_different(const std::vector<int> &addresses)
+    {
+        for (int i = 1; i < addresses.size(); ++i) {
+            if (this->compute_tag(addresses[i - 1]) == this->compute_tag(addresses[i])) {
```

Comment:
But this won't detect the 1st entry equaling the 3rd or 4th or 5th, or the 2nd equaling the 4th, etc.

Reply:
Updated the logic.